File: game/job_queue.py
"""Job queue for batching async Writer generation requests."""

from __future__ import annotations

from collections import deque

from pydantic import BaseModel


class CardGenJob(BaseModel):
    """A single card generation job for the Writer."""

    job_type: str  # "plot" | "event_start" | "event_phase" | "chain" | "info"
    context: dict = {}  # Extra context: plot description, event def, chain tag, etc.


class JobQueue:
    """Accumulates card generation jobs between Writer calls."""

    def __init__(self) -> None:
        self._pending: deque[CardGenJob] = deque()

    def enqueue(self, job: CardGenJob) -> None:
        self._pending.append(job)

    def drain(self) -> list[CardGenJob]:
        """Pop all pending jobs and return them."""
        jobs = list(self._pending)
        self._pending.clear()
        return jobs

    @property
    def has_jobs(self) -> bool:
        return len(self._pending) > 0

    @property
    def count(self) -> int:
        return len(self._pending)

    def has_high_priority(self) -> bool:
        """True if there's a job that should force an early generation."""
        return any(
            j.job_type in ("event_start", "plot")
            for j in self._pending
        )

File: server/internal/game/job_queue.go
package game

import "container/list"

// CardGenJob represents a single card generation job for the Writer
type CardGenJob struct {
	JobType string                 `json:"job_type"` // "plot" | "event_start" | "event_phase" | "chain" | "info"
	Context map[string]interface{} `json:"context"`  // Extra context: plot description, event def, chain tag, etc.
}

// JobQueue accumulates card generation jobs between Writer calls
type JobQueue struct {
	pending *list.List // *CardGenJob
}

// NewJobQueue creates a new job queue
func NewJobQueue() *JobQueue {
	return &JobQueue{
		pending: list.New(),
	}
}

// Enqueue adds a job to the queue
func (jq *JobQueue) Enqueue(job *CardGenJob) {
	jq.pending.PushBack(job)
}

// Drain pops all pending jobs and returns them
func (jq *JobQueue) Drain() []*CardGenJob {
	var jobs []*CardGenJob
	for elem := jq.pending.Front(); elem != nil; elem = elem.Next() {
		jobs = append(jobs, elem.Value.(*CardGenJob))
	}
	jq.pending.Init()
	return jobs
}

// HasJobs returns true if there are pending jobs
func (jq *JobQueue) HasJobs() bool {
	return jq.pending.Len() > 0
}

// Count returns the number of pending jobs
func (jq *JobQueue) Count() int {
	return jq.pending.Len()
}

// HasHighPriority returns true if there's a job that should force early generation
func (jq *JobQueue) HasHighPriority() bool {
	for elem := jq.pending.Front(); elem != nil; elem = elem.Next() {
		job := elem.Value.(*CardGenJob)
		if job.JobType == "event_start" || job.JobType == "plot" {
			return true
		}
	}
	return false
}
